Enable parameters tests

esteve opened the pull request:
Make sure that parameters tests are built and run.

dirk-thomas commented on test_rclcpp/test/test_local_parameters.cpp:
```diff
@@ -39,10 +39,3 @@ TEST(parameters, local_asynchronous) {
-
-int main(int argc, char ** argv)
-{
-  rclcpp::init(argc, argv);
```
Without this custom main `rclcpp::init` is not called anywhere.

esteve replied:
Fixed, added `rclcpp::init` to the tests.

esteve commented:
In https://github.com/ros2/system_tests/commit/faea4a558cd3a15b3a493e790bb0c3ba998eaa1b#diff-a9fbdd47e2e789e3da282a7c222d166dR158 it looks to me that the test will never be built, because the target is not set anywhere. Should that be changed?

dirk-thomas commented:
`ament_add_gtest(target ..)` creates a target with the given name if gtest is available. If gtest is not available it should be mentioned as a warning during configuration but since we always have `gtest_vendor` as a fallback now that should never be the case.

esteve commented:
@dirk-thomas is that a reply to my question or to something in the code of the pull request? If it's the former, I can't seem to find where the target `test_remote_parameters_cpp__${middleware_impl})` is defined.

dirk-thomas commented:
Sorry, I looked at the wrong target `gtest_local_parameters__...`.

Yes, checking for the `test_remote_parameters_cpp__...` `TARGET` first and then creating it with `add_executable` seems wrong. I guess the check should be removed altogether.

esteve commented:
@dirk-thomas no problem :-) I removed the check so that the executables are always built.

esteve commented:
I've added a custom main to the timer tests, updated the PR description to reflect the changes and squashed the commits. Tests pass locally on Linux, these are the CI jobs. They may fail on Windows, since services don't work on Windows (I'm working on that on another branch):

http://ci.ros2.org/job/ros2_batch_ci_windows/378/ (pending, Windows slave is offline)
http://ci.ros2.org/job/ros2_batch_ci_osx/252/
http://ci.ros2.org/job/ros2_batch_ci_linux/322/

dirk-thomas commented:
I have separated the `SKIP_LINKING_MAIN_LIBRARIES` option from the list of source files for better readability in 0f79015d6623a1847181c7ca0ebae346e16c151b.

esteve commented:
@dirk-thomas thanks!

esteve commented:
@dirk-thomas just squashed the branch to include your change.

dirk-thomas commented on test_rclcpp/CMakeLists.txt:
```diff
@@ -135,45 +136,54 @@ if(AMENT_ENABLE_TESTING)
+    ament_target_dependencies(test_parameters_server_cpp__${middleware_impl}
+      "${middleware_impl}"
+      "rclcpp")
+    set(TEST_PARAMETERS_SERVER_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/test_parameters_server_cpp__${middleware_impl}${test_executable_extension}")
```
The variable `test_executable_extension` seems not to be defined. In order to work on Windows I think something like ament/ament_cmake#28 will be necessary to point to the correct location of the `.exe` file.

esteve replied:
There's this snippet in https://github.com/ros2/system_tests/blob/master/test_communication/CMakeLists.txt#L44 to set `test_executable_extension` for the tests in `test_communication`, but seems cumbersome to add it to every `CMakeLists.txt`

How would the approach from ament/ament_cmake#28 work? Using `get_target_property(executable "${target}" LOCATION)` ?

dirk-thomas replied:
`test_executable_extension` is not sufficient. I don't think it works for the test_communication package.

Yes `LOCATION`, and replacing the configuration variable with a generator expression.

esteve replied:
I changed `CMakeLists.txt` to use `get_target_property(...)`, tests pass. If the changes look fine, I'll update `test_communication` to use the same approach too.

esteve replied:
I ended up updating `test_communication` anyway :-)

dirk-thomas commented:
Current CI not passing yet. Moving back to in-progress.

esteve commented:
There's not much work left to do in this PR, the CI just shows that the parameters tests are not actually added to CMake on master.

esteve commented:
I've pushed this branch as `fix-windows-crash` together with the other branches from the same set and the CI job only fails for one test, but I suspect it'll be fixed when Jackie merges the QoS fix:

http://ci.ros2.org/job/ros2_batch_ci_linux/339/

I couldn't get the tests to fail locally, though, despite multiple runs.

The Windows CI job fails because it can't find the GTest headers, I'm working on that:

http://ci.ros2.org/job/ros2_batch_ci_windows/389/

esteve commented:
GTest include dirs fixed.

jacquelinekay commented:
I see a few local test failures (Linux):

```
The following tests FAILED:
	  6 - gtest_local_parameters__rmw_connext_cpp (Failed)
	  7 - parameter_server_cpp__rmw_connext_cpp (Failed)
	 13 - gtest_local_parameters__rmw_connext_dynamic_cpp (Timeout)
	 14 - parameter_server_cpp__rmw_connext_dynamic_cpp (Failed)
	 20 - gtest_local_parameters__rmw_opensplice_cpp (Failed)
	 21 - parameter_server_cpp__rmw_opensplice_cpp (Timeout)
```

I built it with the services QoS fix but they still fail. I'm not sure if it's a problem with my local setup. Also, I wonder if it has to do with merging the fix_future changes, though the jenkins build for that branch was okay.

esteve commented:
The Jenkins build passed becase these tests didn't run on master, the CMake targets were not added appropriately.

esteve commented:
I've run this PR along with the `fix-crash-windows` branches rebased on top of the latest `master` and all tests pass locally on Linux. I'm going to check they run on Windows as well.

jacquelinekay commented:
Are the `fix-crash-windows` branches in a PR somewhere? If they're needed for these tests to pass, they should get merged first. I think I'm seeing the same segfault on my Linux machine in `set_parameters_atomically` that is referenced in ros2/ros2/#109.

esteve commented:
No, they not ready for review yet.

wjwwood commented:
From the meeting:

- Split this pr into one which fixes test finding on Windows, which should be able to be merged directly, and a pr which enables some of the tests which fail on all platforms, which we can put off until we have a fix.

esteve commented:
Spun off to #46 to add the GTest include dir.

dirk-thomas commented on test_communication/CMakeLists.txt:
```diff
@@ -166,8 +155,10 @@ if(AMENT_ENABLE_TESTING)
       endif()
-      set(TEST_PUBLISHER_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/test_publisher_cpp__${middleware_impl1}${test_executable_extension}")
-      set(TEST_SUBSCRIBER_EXECUTABLE "${CMAKE_CURRENT_BINARY_DIR}${test_executable_subfolder}/test_subscriber_cpp__${middleware_impl2}${test_executable_extension}")
+      get_target_property(TEST_PUBLISHER_EXECUTABLE test_publisher_cpp__${middleware_impl1} LOCATION)
```
This needs the same policy block as the other PR to suppress the CMake warning.

esteve replied:
Fixed.

esteve commented:
CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_windows/408/
http://ci.ros2.org/job/ros2_batch_ci_linux/362/
http://ci.ros2.org/job/ros2_batch_ci_osx/272/

The enabled tests pass on Linux and OSX:

http://ci.ros2.org/job/ros2_batch_ci_linux/362/testReport/%28root%29/
http://ci.ros2.org/job/ros2_batch_ci_osx/272/testReport/%28root%29/

Given that tests pass on Linux and OSX, I'd argue that this could be merged, and iterate to improve Windows.

dirk-thomas commented:
The Windows CI job has +124 failing tests. And the failing tests seem to not find the executables: e.g. http://ci.ros2.org/job/ros2_batch_ci_windows/408/testReport/(root)/test_publisher_subscriber_cpp__empty__rmw_connext_cpp/test_publisher_subscriber/

The same test seem to have worked before: http://ci.ros2.org/job/ros2_batch_ci_windows/407/testReport/(root)/test_publisher_subscriber_cpp__empty__rmw_connext_cpp/test_publisher_subscriber/

Therefore I don't think this PR is ready to merged yet.

esteve commented:
@dirk-thomas but if the changes in this PR don't touch the code of those tests, that leads me to believe that they are just flaky on Windows. Merging this PR would just make it easier to iterate, otherwise for every new fix I submit for Windows, I have to  push this branch with yet another name so the CI jobs can catch it.

dirk-thomas commented:
This patch changes the way the executable location is set. And with these new locations a lot of previously succeeding tests fail because the passed executables can not be found at the new location. I don't think there is anything flaky here but the patch is not working as expected.

esteve commented:
@dirk-thomas ok, so I'm going to spin off the changes `get_target_property` changes to a separate branch and just enable the parameters tests.

dirk-thomas commented:
If you look at the generated launch files (e.g. http://ci.ros2.org/job/ros2_batch_ci_windows/ws/workspace/build/test_communication/test_publisher_subscriber_cpp__empty__rmw_connext_cpp.py/*view*/):

    from launch import LaunchDescriptor
    from launch.exit_handler import primary_exit_handler
    from launch.launcher import DefaultLauncher
    
    
    def test_publisher_subscriber():
        ld = LaunchDescriptor()
    
        ld.add_process(
            cmd=['C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/test_communication/$<CONFIGURATION>/test_publisher_cpp__rmw_connext_cpp.exe', 'empty'],
            name='test_publisher',
        )
    
        ld.add_process(
            cmd=['C:/Jenkins/workspace/ros2_batch_ci_windows/workspace/build/test_communication/$<CONFIGURATION>/test_subscriber_cpp__rmw_connext_cpp.exe', 'empty'],
            name='test_subscriber',
            exit_handler=primary_exit_handler,
        )
    
        launcher = DefaultLauncher()
        launcher.add_launch_descriptor(ld)
        rc = launcher.launch()
    
        assert rc == 0, \
            "The launch file failed with exit code '" + str(rc) + "'. " \
            'May be the subscriber did not receive any messages?'
    
    
    if __name__ == '__main__':
        test_publisher_subscriber()

you will see that the approach of this patch (using the generator expression directly) does not work. That is only possible if the path is evaluated by CMake which will take care of replacing the expression with the actual subfolder.

jacquelinekay commented:
On another note: the remote_parameters tests and parameters_server tests hang on my machine, but pass on Jenkins. Is there a step I'm missing here?

esteve commented:
I've removed the `get_target_property` patch and these are new CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_windows/409/
http://ci.ros2.org/job/ros2_batch_ci_osx/273/
http://ci.ros2.org/job/ros2_batch_ci_linux/364/

the timer test fails on OSX and parameter_server_cpp__rmw_connext_dynamic_cpp on Linux. The latter seems to be the what @jacquelinekay has seen locally, my guess is that it could be related to the QoS fix for https://github.com/ros2/rmw_connext/issues/96

dirk-thomas commented:
Since the CI job links themself are not very informative I compared them to current jobs run against master:

The new parameter_server test on Linux (http://ci.ros2.org/job/ros2_batch_ci_linux/364/testReport/(root)/test_rclcpp/parameter_server_cpp__rmw_connext_dynamic_cpp_xunit_missing_result/) seems to timeout. Since it doesn't print anything I don't know what could cause that.

The failing test on OS X (http://ci.ros2.org/job/ros2_batch_ci_osx/273/testReport/(root)/test_time__rmw_connext_cpp/timer_during_wait/) seems to be the flaky one tickets here: ros2/system_tests#36

On Windows it adds 14 new tests and fails 16 tests more than plain master. I can't see what the difference is due to the amount of failing tests.

jacquelinekay commented:
It appears to me that on Windows all of the new parameters test
(local_synchronous, local_asynchronous, remote_paramters,
parameters_server) fail. So the new tests make up the majority of the new
failing tests.

http://ci.ros2.org/job/ros2_batch_ci_windows/409/testReport/%28root%29/parameters/local_synchronous/
http://ci.ros2.org/job/ros2_batch_ci_windows/409/testReport/%28root%29/parameters/local_asynchronous/
http://ci.ros2.org/job/ros2_batch_ci_windows/409/testReport/%28root%29/test_parameters_server_cpp__rmw_connext_cpp/test_remote_parameters/
http://ci.ros2.org/job/ros2_batch_ci_windows/409/testReport/projectrootC_.Jenkins.workspace.ros2_batch_ci_windows.workspace.src.ros2.system_tests/test_rclcpp/parameter_server_cpp__rmw_opensplice_cpp/

On Wed, Sep 16, 2015 at 2:48 PM, Dirk Thomas <notifications@github.com>
wrote:

> Since the CI job links themself are not very informative I compared them
> to current jobs run against master:
>
> The new parameter_server test on Linux (
> http://ci.ros2.org/job/ros2_batch_ci_linux/364/testReport/(root)/test_rclcpp/parameter_server_cpp__rmw_connext_dynamic_cpp_xunit_missing_result/)
> seems to timeout. Since it doesn't print anything I don't know what could
> cause that.
>
> The failing test on OS X (
> http://ci.ros2.org/job/ros2_batch_ci_osx/273/testReport/(root)/test_time__rmw_connext_cpp/timer_during_wait/)
> seems to be the flaky one tickets here: ros2/system_tests#36
> <https://github.com/ros2/system_tests/issues/36>
>
> On Windows it adds 14 new tests and fails 16 tests more than plain master.
> I can't see what the difference is due to the amount of failing tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/system_tests/pull/44#issuecomment-140903981>.
>

dirk-thomas commented:
Especially the tests failing with `The system cannot find the file specified` should be straight forward to fix.

esteve commented:
@dirk-thomas when the `parameter_server_cpp__rmw_connext_dynamic_cpp` test is run manually, the client gets stuck waiting for a response from the server, but never occurs. The branches for https://github.com/ros2/rmw_connext/issues/96 seem to fix it.

@jacquelinekay Windows is currently a bit of a mess, many other tests fail too. I'll fix the ones that can't find the executables, but the rest will require more work.

esteve commented:
I've replaced `get_target_property` logic with setting the path to the binaries manually and the test now can find the executables, though some still fail on Windows.

esteve commented:
CI jobs:

http://ci.ros2.org/job/ros2_batch_ci_linux/430/
http://ci.ros2.org/job/ros2_batch_ci_osx/340/
http://ci.ros2.org/job/ros2_batch_ci_windows/482/

esteve commented:
This now passes on Windows locally, let's see if it does on the farm:

http://ci.ros2.org/job/ros2_batch_ci_linux/453/
http://ci.ros2.org/job/ros2_batch_ci_osx/365/
http://ci.ros2.org/job/ros2_batch_ci_windows/510/

esteve commented:
CI after rebasing against master:

http://ci.ros2.org/job/ros2_batch_ci_linux/456/
http://ci.ros2.org/job/ros2_batch_ci_osx/368/
http://ci.ros2.org/job/ros2_batch_ci_windows/512/

previous run failed on Windows, but I can't reproduce it locally after multiple runs.

esteve commented:
CI after rebasing against master:

http://ci.ros2.org/job/ros2_batch_ci_linux/457/
http://ci.ros2.org/job/ros2_batch_ci_osx/369/
http://ci.ros2.org/job/ros2_batch_ci_windows/513/

previous run failed on Windows, but I can't reproduce it locally after multiple runs.

esteve commented:
So it seems the failing tests on Windows are only for RTI Connext, this might be related to https://github.com/ros2/rmw_connext/issues/31

The OpenSplice versions of the tests pass on Windows:

http://ci.ros2.org/job/ros2_batch_ci_windows/513/testReport/%28root%29/test_local_parameters__rmw_opensplice_cpp/
http://ci.ros2.org/job/ros2_batch_ci_windows/513/testReport/%28root%29/test_parameters_server_cpp__rmw_opensplice_cpp/

jacquelinekay commented:
+1 with squash. I am also convinced that the failures on Windows are due to the requester/replier issue.

esteve commented:
Squashed and merging now.
